Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+        detail::PointwiseMaterialInteraction interaction(surface, state,
+                                                         stepper);
+        // Evaluate the material properties
+        if (interaction.evaluateMaterialProperties(state, updateStage)) {
+          // Surface has material at this stage
+          hasMaterial = true;
+
+          // Evaluate the material effects
+          interaction.evaluatePointwiseMaterialInteraction(multipleScattering,
+                                                           energyLoss);
+
+          // Screen out material effects info
+          ACTS_VERBOSE("Material effects on surface: "
+                       << surface->geoID()
+                       << " at update stage: " << updateStage << " are :");
+          ACTS_VERBOSE("eLoss = "
+                       << interaction.Eloss << ", "
+                       << "variancePhi = " << interaction.variancePhi << ", "
+                       << "varianceTheta = " << interaction.varianceTheta
+                       << ", "
+                       << "varianceQoverP = " << interaction.varianceQoverP);
+
+          // Update the state and stepper with material effects
+          interaction.updateState(state, stepper);
+        }
+      }
+
+      if (not hasMaterial) {
+        // Screen out message
+        ACTS_VERBOSE("No material effects on surface: " << surface->geoID()
+                                                        << " at update stage: "
+                                                        << updateStage);
+      }
+    }
+
+    /// @brief Kalman actor operation : finalize
+    ///
+    /// @tparam propagator_state_t is the type of Propagagor state
+    /// @tparam stepper_t Type of the stepper
+    ///
+    /// @param state is the mutable propagator state object
+    /// @param stepper The stepper in use
+    /// @param result is the mutable result state object
+    template <typename propagator_state_t, typename stepper_t>
+    Result<void> finalize(propagator_state_t& state, const stepper_t& stepper,
+                          result_type& result) const {
+      // Get the index of measurement states;
+      std::vector<size_t> measurementIndices;
+      auto lastState = result.fittedStates.getTrackState(result.currentTip);
+      if (lastState.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        measurementIndices.push_back(result.currentTip);
+      }
+      // Count track states to be smoothed
+      size_t nStates = 0;
+      result.fittedStates.applyBackwards(result.currentTip, [&](auto st) {
+        // Smoothing will start from the last measurement state
+        if (measurementIndices.empty()) {
+          // No smoothed parameter for the last few non-measurment states
+          st.data().ismoothed = detail_lt::IndexData::kInvalid;
+        } else {
+          nStates++;
+        }
+        size_t iprevious = st.previous();
+        if (iprevious != Acts::detail_lt::IndexData::kInvalid) {
```

Comment:
Maybe I should add a `hasPrevious()` method to the TSproxy.

Reply:
However, could you not skip this by just pushing back the result of the flag test for the current index, rather than the previous one? So you get rid of the special case for `result.currentTip` outside of the lambda, and inside the lambda just push back for the current TS.

Reply:
What a shame that I don't realize there is the `index()` method! This part of code has been much simplified now. And yes, the `hasPrevious()` method could be very useful.